### DISCUSSION TITLE
Get WM info into eww

### DISCUSSION THREAD

**Discussion by VuiMuich** [elkowar/eww#167] [Q&A] [answered]
Hi, 
first of all thanks, eww is absolutely awesome!

I just have one thing I couldn't figure out yet:
how do I get info from the WM like window-title of the focused window, status of the workspaces etc.?

I am using `leftwm` and for some bars pipes it's state to the launcher of the bar. For example with xmobar it looks something like this:
```
leftwm-state -w $index -t $SCRIPTPATH/template.liquid | xmobar -p "$size" $SCRIPTPATH/xmobar-conf │ ig.hs &
```
Or with polybar it launches the command from within `polybar.config` so I tried:
```xml
<script-var name="leftwm-state" interval="55ms">leftwm-state -w 0 -t ~/.config/leftwm/themes/current/template.liquid</script-var>
```
But the variable empty.

Exemplary the ouput of leftwm-stat  looks like this:
```
> leftwm-state -w 0 -t ~/.config/leftwm/themes/current/template.liquid
%{F#554d8b}%{B#232131}  Home  %{B-}%{F-}%{F#554d8b}%{B#232131}  Web  %{B-}%{F-}%{F#d8d8d8}%{B#554d8b}  Code  %{B-}%{F-}%{F#554d8b}%{B#232131}  Misc  %{B-}%{F-}%{F#554d8b}%{B#232131}  5  %{B-}%{F-}%{F#999999}%{B#232131} · %{B-}%{F-}%{F#999999}%{B#232131} · %{B-}%{F-}%{F#999999}%{B#232131} · %{B-}%{F-}%{F#999999}%{B#232131} · %{B-}%{F-}%{r}%{F#554d8b}%{B#232131}  leftwm-state -w 0 -t ~/.config/leftwm/themes/current/template.liquid /home/vuimuich/.config/leftwm/themes/floaters  { %{B-}%{F-}
```
Bonus will be to parse this, so the workspaces will get assigned to buttons, but at first I would be happy to get leftwm-state at least read..

**Comment by elkowar** [comment]
I'm assuming that that command continuously prints output, whenever the state changes, right? In that case, you don't want to set an interval, as that reruns the command every 55ms. The likelyhood of anything happening within those 55ms is rather low, thus probably you're just never getting any output for long enough to register.

Try just removing `interval="55ms"` from your script-var and see if that helps

**Reply by VuiMuich** [reply]
Hey, thanks for that quick answer!
This gets me bit further:
I figured how to modify the `template.liquid` file so I basically get correct xml code for the workspace buttons, the output is now like this:
```xml
<button class="ws-button-busy" onclick="wmctrl -s Home">  Home  </button><button class="ws-button-busy" onclick="wmctrl -s Web">  Web  </button><button class="ws-button-mine" onclick="wmctrl -s Code">  Code  </button><button class="ws-button-busy" onclick="wmctrl -s Misc">  Misc  </button><button class="ws-button-busy" onclick="wmctrl -s 5">  5  </button><button class="ws-button-busy" onclick="wmctrl -s 6"> · </button><button class="ws-button-busy" onclick="wmctrl -s 7"> · </button><button class="ws-button-busy" onclick="wmctrl -s 8"> · </button><button class="ws-button-busy" onclick="wmctrl -s 9"> · </button>
```
I can paste this in the `workspace` wiget and it works so now it just needs to get processed and infused correctly

**Reply by elkowar** [reply]
well then, just use a `<literal>` to put the generated XML into the widget!

**Reply by VuiMuich** [reply]
hmm, somehow `eww state` appears to be frozen.
I tried `eww update` and `eww reload` but the variables don't change.
`eww logs` just states `Reloaded config successfully` :confused:

edit: well not all of the varialbes. e.g. batteryis changing its value..

gonna try `<literal>` (note to self: read documentation more carfully)

**Reply by elkowar** [reply]
which variables _should_ be changing?

**Reply by VuiMuich** [reply]
Just to get it right how to set up the `<literal>`:
1. i have a `<script-var name="wmstate">` with my command/external script
2. I put `<literal>{{wmstate]}</literal> in a `<box>` right?

hmm, now I get actualy errors in the log:
```thread 'main' panicked at 'failed printing to stdout: Broken pipe (os error 32)', library/std/src/io/stdio.rs:940:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

I guess it was just me, who got confused and thought `eww state` was frozen..

**Reply by elkowar** [reply]
you put `<literal content="{{wmstate}}"/>`

**Reply by VuiMuich** [reply]
Ohh.. will try this, in the evening.

**Reply by VuiMuich** [reply]
Ok, slowly I feel I am doing any possible stupid mistake.
With the literal being as above, now I get 
`Faailed to parse eww xml literal: ParserError(UnknownToken(TextPos { row: 1, col: 74 }))`
Which is exactly the opening '<' of the second button.
Do I need to escape some characters for the parser to get this right?

**Reply by elkowar** [reply]
Well, no, that just means your script is outputting invalid xml ig

**Reply by VuiMuich** [reply]
But when I take the output of `eww state` and copy the variable `wmstate` and paste it in place of `<literal content={{wmstate}}/>` the buttons are drawn (of course it's static except `hover` and `onclick` then).

> For reference again: this is what the `<script-var name="wmstate">` contains:
> ```xml
> <button class="ws-button-busy" onclick="wmctrl -s Home">  Home  </button><button class="ws-button-busy" onclick="wmctrl -s Web">  Web  </button><button class="ws-button-mine" onclick="wmctrl -s Code">  Code  </button><button class="ws-button-busy" onclick="wmctrl -s Misc">  Misc  </button><button class="ws-button-busy" onclick="wmctrl -s 5">  5  </button><button class="ws-button" onclick="wmctrl -s 6"> · </button><button class="ws-button" onclick="wmctrl -s 7"> · </button><button class="ws-button" onclick="wmctrl -s 8"> · </button><button class="ws-button" onclick="wmctrl -s 9"> · </button>
> ```

I will try to put this in a static variable, and also run eww with RUST_TRACEBAK again.

**Reply by elkowar** [reply]
a literal needs to take a _single_ element, so you'll need to wrap your buttons in a `<box>` or something like that.

**Reply by VuiMuich** [reply]
That did the trick! Thanks!
I will compile the infor for the proper solution into one comment later.

(Just one minor thing that puzzles me is how to space buttons properly, but I guess this is just getting more fluent in css and tag parameters:
![2021-05-14-211637_537x34_scrot](https://user-images.githubusercontent.com/4779365/118318486-bf0ba580-b4f9-11eb-9b9b-ff8ae2afd3fe.png))

**Comment by VuiMuich** [comment, marked as answer]
This compiles the key insights of the thread in the other suggestion in order to have a solution in a glance:

- In order get dynamically generated cal code from a script use the `<literal>` tag
- The literal tag receives the xml code with it `content` parameter like this `<literal content={{variable-Name}}/>`
- the literal only takes one element, if more elements are needed put them in a `<box>`
